File: src/main/java/club/libridge/libridgebackend/core/Contract.java
package club.libridge.libridgebackend.core;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NonNull;

@AllArgsConstructor
public final class Contract {
    @NonNull
    private final OddTricks oddTricks;
    @Getter
    @NonNull
    private final Strain strain;
    @NonNull
    private final PenaltyStatus penaltyStatus;
    @Getter
    private final boolean vulnerable;

    public int getLevel() {
        return this.oddTricks.getLevel();
    }

    public boolean isDoubled() {
        return PenaltyStatus.DOUBLED == this.penaltyStatus;
    }

    public boolean isRedoubled() {
        return PenaltyStatus.REDOUBLED == this.penaltyStatus;
    }

    @Override
    public String toString() {
        String response = this.getLevel() + this.getStrain().getSymbol();
        if (this.isDoubled()) {
            response += "X";
        } else if (this.isRedoubled()) {
            response += "XX";
        }
        return response;
    }

}

File: src/main/scala/scalabridge/Contract.scala
package scalabridge

sealed abstract class Contract {
  val isAllPass: Boolean
}
object Contract {
  def apply(
      oddTricks: OddTricks,
      strain: Strain,
      penaltyStatus: PenaltyStatus
  ) = DefaultContract(oddTricks, strain, penaltyStatus)
}

case class DefaultContract(
    oddTricks: OddTricks,
    strain: Strain,
    penaltyStatus: PenaltyStatus
) extends Contract {

  override val isAllPass: Boolean = false

  def getLevel = oddTricks.getLevel
  def isDoubled = PenaltyStatus.DOUBLED == penaltyStatus
  def isRedoubled = PenaltyStatus.REDOUBLED == penaltyStatus
  override def toString(): String = {
    val response = s"${this.getLevel.toString()}${this.strain.getSymbol}";
    if (this.isDoubled) s"${response}X";
    else if (this.isRedoubled) s"${response}XX";
    else response;
  }
  def getStrain = this.strain
}

case object AllPassContract extends Contract {
  override val isAllPass: Boolean = true
  override def toString(): String = "ALLPASS"
}
